breacharbiter.go

```diff
@@ -907,61 +989,101 @@ func (b *breachArbiter) createJusticeTx(
-	sweepTx := wire.NewMsgTx(1)
-	sweepTx.AddTxIn(&wire.TxIn{
-		PreviousOutPoint: *closeInfo.SelfOutPoint,
-	})
-	sweepTx.AddTxOut(&wire.TxOut{
-		PkScript: sweepPkScript,
-		Value:    int64(sweepAmt),
+	feePerWeight := b.estimator.EstimateFeePerWeight(1)
+	txFee := btcutil.Amount(txWeight * feePerWeight)
+
+	sweepAmt := int64(totalAmt - txFee)
+
+	// With the fee calculated, we can now create the transaction using the
+	// information gathered above and the provided retribution information.
+	var txn = wire.NewMsgTx(2)
+
+	// We begin by adding the output to which our funds will be deposited.
+	txn.AddTxOut(&wire.TxOut{
+		PkScript: pkScript,
+		Value:    sweepAmt,
 	})
 
-	// Next, we'll generate the signature required to satisfy the p2wkh
-	// witness program.
-	signDesc := closeInfo.SelfOutputSignDesc
-	signDesc.SigHashes = txscript.NewTxSigHashes(sweepTx)
-	signDesc.InputIndex = 0
-	sweepSig, err := b.wallet.Cfg.Signer.SignOutputRaw(sweepTx, signDesc)
-	if err != nil {
-		return nil, err
+	// Next, we add all of the spendable outputs as inputs to the
+	// transaction.
+	for _, input := range inputs {
+		txn.AddTxIn(&wire.TxIn{
+			PreviousOutPoint: *input.OutPoint(),
+		})
 	}
 
-	// Finally, we'll manually craft the witness. The witness here is the
-	// exact same as a regular p2wkh witness, but we'll need to ensure that
-	// we use the tweaked public key as the last item in the witness stack
-	// which was originally used to created the pkScript we're spending.
-	witness := make([][]byte, 2)
-	witness[0] = append(sweepSig, byte(txscript.SigHashAll))
-	witness[1] = lnwallet.TweakPubKeyWithTweak(
-		signDesc.PubKey, signDesc.SingleTweak,
-	).SerializeCompressed()
+	// Create a sighash cache to improve the performance of hashing and
+	// signing SigHashAll inputs.
+	hashCache := txscript.NewTxSigHashes(txn)
+
+	// Create a closure that encapsulates the process of initializing a
+	// particular output's witness generation function, computing the
+	// witness, and attaching it to the transaction. This function accepts
+	// an integer index representing the intended txin index, and the
+	// breached output from which it will spend.
+	addWitness := func(idx int, so SpendableOutput) error {
+		// First, we construct a valid witness for this outpoint and
+		// transaction using the SpendableOutput's witness generation
+		// function.
+		witness, err := so.BuildWitness(
+			b.wallet.Cfg.Signer, txn, hashCache, idx,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Then, we add the witness to the transaction at the
+		// appropriate txin index.
+		txn.TxIn[idx].Witness = witness
 
-	sweepTx.TxIn[0].Witness = witness
+		return nil
+	}
 
-	brarLog.Infof("Sweeping commitment output with: %v", spew.Sdump(sweepTx))
+	// Finally, generate a witness for each output and attach it to the
+	// transaction.
+	for i, input := range inputs {
+		if err := addWitness(i, input); err != nil {
+			return nil, err
+		}
+	}
 
-	return sweepTx, nil
+	return txn, nil
```

Comment:
Before returning transaction for singing, we should ensure that the transaction itself is sane. We can use `blockchain.CheckTransactionSanity` for this purpose. This'll catch things like a negative output due to a very high estimated fee. 

In general whenever crafting transactions like this, we'll need to ensure that we don't try to create negative valued outputs, the `CheckTransactionSanity` function can do this job for us.

Reply:
Didn't realize this was available, makes a lot of sense! :)